Coalesce

vasilyrud opened the pull request:
First thing: 
I have an issue in that my .gitignore doesn't actually ignore things properly. As you can see, after I did `make clean` the deletions became part of this commit, which is not nice. Not sure how to fix that.

Second:
[http://68.media.tumblr.com/tumblr_lwx8xoX06z1qj4b9to1_500.gif](http://68.media.tumblr.com/tumblr_lwx8xoX06z1qj4b9to1_500.gif)

Third - changes made:
- Added the macros `PREV_BLOCK_IN_MEM` and `NEXT_BLOCK_IN_MEM` both of which return to the header of the previous and next block respectively. Made one combined test for both.
- Made a minor change to the test for `FTRP`, the total size was off by `HDR_FTR_SIZE`.
- Added coalesce and added one test function which tests for each of the four possible cases of coalescing.

@quanvuong

vasilyrud committed:
start coalesce

rebase

rebase

fix comment

start coalesce test

add prev/next in mem macros and coalesce

change test sizes

improve coalesce tests, add function

quanvuong commented on mm.c:
```diff
@@ -59,26 +59,32 @@ team_t team = {
 #define GET_TOTAL_SIZE(p) (GET_SIZE(p) + HDR_FTR_SIZE)
 
+// Given pointer to current block, return pointer to header of previous block
+#define PREV_BLOCK_IN_MEM(header_p) ((char **)(header_p) - GET_SIZE((char **)(header_p) - 1) - HDR_FTR_SIZE)
```
can we get clearer name here to indicate if next and prev here implies ordering in the heap or in the free list ?

vasilyrud replied:
I tried to achieve that with the `MEM` ending. If I change it to `HEAP` will it be clearer?

quanvuong replied:
yup, because they're all in memory

quanvuong commented on mm.c:
```diff
@@ -118,9 +124,36 @@ static size_t find_free_list_index(size_t words) {
+	if (prev_status == TAKEN && next_status == TAKEN) {
+		return bp;
+	} else if (prev_status == TAKEN && next_status == FREE) {
+		new_size += GET_SIZE(next_block) + HDR_FTR_SIZE;
```
can you use the macro GET_TOTAL_SIZE here ?

vasilyrud replied:
I forgot about that macro, will do.

quanvuong commented on mm.c:
```diff
@@ -118,9 +124,36 @@ static size_t find_free_list_index(size_t words) {
+	if (prev_status == TAKEN && next_status == TAKEN) {
+		return bp;
+	} else if (prev_status == TAKEN && next_status == FREE) {
+		new_size += GET_SIZE(next_block) + HDR_FTR_SIZE;
```
Do we also need to remove the current block and the block that it's coalescing with from their respective free lists ?

vasilyrud replied:
The current block, no, because coalesce is only called once a TAKEN block becomes FREE.

The prev/next blocks, you are right about. We need to make sure that the block is taken out of the free list. 

Do you think it will be a good idea to use `alloc_free_block` to achieve this task? The function does exactly that (removes the block from the free list) and then `coalesce` changes the TAKEN status to FREE.

This will reduce code redundancy, and also that way until `alloc_free_block` is implemented, `coalesce` will still work.

What do you think?

quanvuong replied:
alloc_free_block doesn't seem clearly enough. can we make the name more explicitly ? (e.g. remove_block_from_free_list)

quanvuong commented on mm.c:
```diff
@@ -346,6 +379,103 @@ static void test_FTRP()
+	new_size = curr_size + next_size + HDR_FTR_SIZE;
+	assert(GET_SIZE(p_curr) == new_size);
+	assert(GET_SIZE(FTRP(p_curr)) == new_size);
+	assert(GET_STATUS(p_prev) == TAKEN);
```
can you also check that the previous block is not modified here ?

vasilyrud replied:
Good idea.

quanvuong commented on mm.c:
```diff
@@ -346,6 +379,103 @@ static void test_FTRP()
+	coalesce(p_curr);
+	new_size = prev_size + curr_size + next_size + 2*HDR_FTR_SIZE;
+	assert(GET_SIZE(p_prev) == new_size);
+	assert(GET_SIZE(FTRP(p_prev)) == new_size);
```
can you check that the status of the newly coalesced block here ?

vasilyrud replied:
Good idea.

quanvuong reviewed (commented):
http://giphy.com/gifs/studiosoriginals-wrestling-josh-freydkis-3o7TKHonfdKgUrcl9K

vasilyrud commented:
Ok, I rebased onto master and got the changes implemented.

Instead of changing the name of the function `alloc_free_block` I added a new function `remove_block_from_free_list` that will be used both by `alloc_free_block` and by `coalesce`. This way coalesce will not have to run through the hole of `alloc_free_block` which does some things unrelated to `coalesce`.

quanvuong commented on mm.c:
```diff
@@ -125,9 +133,40 @@ static size_t find_free_list_index(size_t words) {
+	if (prev_status == TAKEN && next_status == TAKEN) {
+		return bp;
+	} else if (prev_status == TAKEN && next_status == FREE) {
+		remove_block_from_free_list(next_block);
```
don't we have to remove the current block from its free list too ?

vasilyrud replied:
As per the description in the comment above the function: "Coalesce is only called when a taken block is freed, before the free block is placed into the free list."

Thus, no.

quanvuong replied:
yeah you're right. done.

quanvuong commented on mm.c:
```diff
@@ -366,6 +417,111 @@ static void test_MAIN_FREE_LIST_INIT()
+	printf("Test passed.\n\n");
+}
+
+static void test_coalesce() {
```
In the interest of bug free code, we should also test that the blocks that are being coalesced have been removed from their respective free blocks. 

This means, we need functions to insert free block, remove free block, only then we can safely test coalesce. I suggest that we do those first, before finishing this PR.

I can take the insert function, do you want to take the remove function. WDYT ?

quanvuong replied:
Quan's writing the two functions place_block_into_free_list and remove_block_from_free_list

quanvuong reviewed (commented):
http://giphy.com/gifs/studiosoriginals-wrestling-josh-freydkis-3o7TKpAPh1MAN05xfy

vasilyrud commented:
This PR is outdated, not to be merged.

quanvuong commented:
can't we savage the coalesce function ?

quanvuong commented:
oh nvm, saw that you have another PR for coalesce
